File: src/main/java/com/zippeykeys/praisebe/item/block/ItemBlockEnum.java
package com.zippeykeys.praisebe.item.block;

import com.zippeykeys.praisebe.block.base.BlockEnum;
import com.zippeykeys.praisebe.util.Localize;
import com.zippeykeys.praisebe.util.MathUtil;

import lombok.val;
import net.minecraft.item.ItemBlock;
import net.minecraft.item.ItemStack;

public class ItemBlockEnum<T extends Enum<T> & Localize> extends ItemBlock {

    public ItemBlockEnum(BlockEnum<T> blockIn) {
        super(blockIn);

        if (blockIn.getValues().length != 0)
            setHasSubtypes(true);
    }

    @SuppressWarnings("unchecked")
    @Override
    public BlockEnum<T> getBlock() {
        return (BlockEnum<T>) super.getBlock();
    }

    @Override
    public String getUnlocalizedName(ItemStack stack) {
        val block = getBlock();
        return block.getUnlocalizedName()
                + block.getValues()[MathUtil.clamp(stack.getItemDamage(), 0, getBlock().getValues().length - 1)]
                        .getName();
    }

    @Override
    public int getMetadata(int damage) {
        return damage;
    }
}

File: src/main/kotlin/com/zippeykeys/praisebe/item/block/ItemBlockEnum.kt
package com.zippeykeys.praisebe.item.block

import com.zippeykeys.praisebe.block.base.BlockEnum
import com.zippeykeys.praisebe.iface.ILocalize
import net.minecraft.item.ItemStack
import net.minecraft.util.math.MathHelper.clamp


class ItemBlockEnum<T>(blockIn: BlockEnum<T>) : PBItemBlock(blockIn) where T : Enum<T>, T : ILocalize {
    init {
        if (blockIn.values.isNotEmpty()) {
            setHasSubtypes(true)
        }
    }

    @Suppress("UNCHECKED_CAST")
    override fun getBlock(): BlockEnum<T> {
        return super.getBlock() as BlockEnum<T>
    }

    override fun getUnlocalizedName(stack: ItemStack): String {
        val block = getBlock()
        return block.unlocalizedName + block.values[clamp(stack.itemDamage, 0, getBlock().values.size - 1)]
                .getName()
    }

    override fun getMetadata(damage: Int): Int {
        return damage
    }
}